File: v2_wsl/cc-communicate/scripts/lib/proc.js
'use strict';
// Cross-platform process introspection.
//   - Windows: PowerShell + CIM
//   - Linux:   /proc
//   - macOS:   `ps -o etime`     (best-effort)
const { execFileSync } = require('child_process');
const fs = require('fs');

// A process is "the claude binary" iff its cmdline's FIRST token (the
// executable) is/ends with `claude` or `claude.exe`. v0.1 tested the FULL
// cmdline for "claude" and excluded cmdlines containing "cc-communicate" - but
// spawn/evoke prompts contain "cc-communicate", so a spawned CC's claude parent
// was rejected and the hook fell back to a fragile ppid guess (v2.2 Amd1 /
// BUG-1). Checking only the first token avoids the prompt text: our scripts run
// as `python ...`/`node ...`, shells as `cmd`/`bash`/`tmux` - none have claude
// as the first token.
function isClaudeCmd(cmd) {
  if (!cmd) return false;
  const first = cmd.split(/\s+/)[0];
  return /(^|\/|\\)claude(\.exe)?$/i.test(first);
}

/* ---- Windows ---------------------------------------------------------- */
function procTableWindows() {
  const ps =
    "Get-CimInstance Win32_Process | Select-Object ProcessId,ParentProcessId,CommandLine," +
    "@{N='Created';E={ if ($_.CreationDate) { $_.CreationDate.ToString('o') } else { '' } }} | " +
    "ConvertTo-Json -Compress";
  const out = execFileSync('powershell',
    ['-NoProfile', '-NonInteractive', '-Command', ps],
    { encoding: 'utf8', maxBuffer: 128 * 1024 * 1024, windowsHide: true });
  let arr = JSON.parse(out);
  if (!Array.isArray(arr)) arr = [arr];
  const map = new Map();
  for (const p of arr) {
    map.set(p.ProcessId, { ppid: p.ParentProcessId, cmd: p.CommandLine || '', start: p.Created || null });
  }
  return map;
}

/* ---- Linux (/proc) ---------------------------------------------------- */
function procTableLinux() {
  const map = new Map();
  let btime = 0;
  try {
    const stat = fs.readFileSync('/proc/stat', 'utf8');
    const m = stat.match(/btime\s+(\d+)/);
    if (m) btime = parseInt(m[1], 10);
  } catch (_) {}
  const hz = 100;
  for (const entry of fs.readdirSync('/proc')) {
    if (!/^\d+$/.test(entry)) continue;
    const pid = parseInt(entry, 10);
    let ppid = 0, cmd = '', starttime = 0;
    try {
      const s = fs.readFileSync(`/proc/${pid}/stat`, 'utf8');
      const closeParen = s.lastIndexOf(')');
      const pre = s.slice(0, closeParen + 1);
      const rest = s.slice(closeParen + 2).split(/\s+/);
      const cmdMatch = pre.match(/\((.*)\)$/);
      cmd = cmdMatch ? cmdMatch[1] : '';
      ppid = parseInt(rest[1], 10);
      starttime = parseInt(rest[19], 10);
    } catch (_) { continue; }
    try { cmd = fs.readFileSync(`/proc/${pid}/cmdline`, 'utf8').replace(/\0/g, ' ').trim() || cmd; } catch (_) {}
    const start = (btime && starttime) ? new Date((btime + starttime / hz) * 1000).toISOString() : null;
    map.set(pid, { ppid, cmd, start });
  }
  return map;
}

/* ---- macOS (ps) ------------------------------------------------------- */
function parseEtimeToSec(e) {
  const m = e.trim().match(/^(?:(\d+)-)?(?:(\d+):)?(\d+):(\d+)$/);
  if (!m) return null;
  const d = m[1] ? +m[1] : 0, h = m[2] ? +m[2] : 0, mi = +m[3], s = +m[4];
  return d * 86400 + h * 3600 + mi * 60 + s;
}
function procTableMac() {
  const out = execFileSync('ps', ['-eo', 'pid=,ppid=,etime=,command='], { encoding: 'utf8' });
  const map = new Map();
  const now = Date.now();
  for (const line of out.split('\n')) {
    if (!line.trim()) continue;
    const m = line.trim().match(/^(\d+)\s+(\d+)\s+(\S+)\s+(.*)$/);
    if (!m) continue;
    const pid = +m[1], ppid = +m[2], etime = m[3], cmd = m[4];
    const sec = parseEtimeToSec(etime);
    const start = sec != null ? new Date(now - sec * 1000).toISOString() : null;
    map.set(pid, { ppid, cmd, start });
  }
  return map;
}

function getProcTable() {
  if (process.platform === 'win32') return procTableWindows();
  if (process.platform === 'linux')  return procTableLinux();
  return procTableMac();
}

// Walk up from selfPid; return the nearest ancestor that is the claude binary.
function resolveClaude(selfPid) {
  let map;
  try { map = getProcTable(); }
  catch (e) { return { pid: process.ppid, start: null, chain: [], error: e.message }; }

  const start = selfPid || process.pid;
  const chain = [];
  const seen = new Set();
  let cur = start, guard = 0;
  while (cur && map.has(cur) && guard++ < 64 && !seen.has(cur)) {
    seen.add(cur);
    const info = map.get(cur);
    chain.push({ pid: cur, cmd: (info.cmd || '').slice(0, 160) });
    if (cur !== start && isClaudeCmd(info.cmd)) {
      return { pid: cur, start: info.start, chain };
    }
    cur = info.ppid;
  }
  const fb = map.get(process.pid) ? map.get(process.pid).ppid : process.ppid;
  return { pid: fb, start: map.has(fb) ? map.get(fb).start : null, chain };
}

module.exports = { getProcTable, resolveClaude, liveProcs, isClaudeCmd };

File: v2_win/cc-communicate/server/proc.py
"""Process introspection for the cc-communicate upper layer.

Frozen-equivalent of the parts of scripts/lib/proc.js that the upper layer
uses. The upper layer reads events that already carry pid + start_time (for
liveness), AND resolves the claude ancestor (for my_session_id - the MCP
server walks up from its own pid to find the calling CC's claude binary).

Uses psutil (cross-platform). On Windows, psutil's create_time() reads the same
Win32 FILETIME that proc.js's CIM branch reads, so epoch values agree across
the two layers for the same process.
"""
from __future__ import annotations

import re
from datetime import datetime

import psutil

_FRACT_TRAIL = re.compile(r"(\.\d{6})\d+")


def live_procs() -> dict[int, float]:
    out: dict[int, float] = {}
    try:
        for p in psutil.process_iter(["pid", "create_time"]):
            ct = p.info.get("create_time")
            if ct is not None:
                out[p.info["pid"]] = float(ct)
    except psutil.Error:
        pass
    return out


def proc_start_time(pid: int) -> float | None:
    try:
        return float(psutil.Process(pid).create_time())
    except psutil.NoSuchProcess:
        return None
    except psutil.AccessDenied:
        return None
    except psutil.Error:
        return None


def pid_matches(pid, recorded):
    """None=unknown/unset, False=dead, True=alive. The shared liveness rule
    (was check_alive._match / session_by_pid._pid_live): the pid must exist
    AND its start time must match the recorded one within 1s (rejects pid
    reuse)."""
    if pid is None or recorded is None:
        return None
    current = proc_start_time(pid)
    if current is None:
        return False
    return abs(current - float(recorded)) <= 1.0


def parse_start_time(s) -> float | None:
    if not s:
        return None
    try:
        t = s.strip()
        t = _FRACT_TRAIL.sub(lambda m: m.group(1), t, count=1)
        if t.endswith("Z"):
            t = t[:-1] + "+00:00"
        return datetime.fromisoformat(t).timestamp()
    except (ValueError, TypeError, AttributeError):
        # AttributeError: non-str truthy input (e.g. a float epoch) has no
        # .strip() - treat as unparseable instead of crashing (Wave 1
        # hardening; today's producers write ISO strings or null only).
        return None


def resolve_claude(self_pid: int):
    """Walk up the process tree from self_pid to find the claude binary ancestor.
    Returns (pid, start_time_epoch) or (None, None).

    Identification is by process NAME (claude / claude.exe), NOT by cmdline
    substring. v0.1 matched "claude" in cmdline and skipped cmdlines containing
    "cc-communicate" - but the spawn/evoke prompts literally contain
    "cc-communicate", so a spawned CC's claude parent was rejected and
    my_session_id failed (v2.2 Amd1 / BUG-1). Our own scripts run as python/node
    (name python/node); intermediate shells are cmd/bash/tmux - none named
    claude - so a name check cleanly distinguishes the real claude binary from
    processes whose cmdline merely references claude (e.g. the prompt text)."""
    try:
        p = psutil.Process(self_pid)
        for parent in p.parents():
            try:
                name = (parent.name() or "").lower()
            except (psutil.AccessDenied, psutil.NoSuchProcess):
                continue
            if name not in ("claude", "claude.exe"):
                continue
            return parent.pid, parent.create_time()
    except (psutil.NoSuchProcess, psutil.AccessDenied):
        pass
    return None, None


def claude_binary_path(self_pid: int):
    """Absolute path to the claude binary ancestor's executable, or None.

    On WSL `which claude` returns the Windows version (C13); the kernel detects
    its own claude ancestor's exe() at init so spawn.py uses the full Linux
    path. On Windows this is unnecessary (claude is on PATH)."""
    pid, _ = resolve_claude(self_pid)
    if pid is None:
        return None
    try:
        return psutil.Process(pid).exe()
    except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.Error):
        return None
